File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/MemberMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.*
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONException
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import com.twilio.util.ErrorInfo
import android.util.Log

object MemberMethods {
    fun getChannel(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioConversationsPlugin.chatClient?.getConversation(channelSid, object : CallbackListener<Conversation> {
            override fun onSuccess(channel: Conversation) {
                Log.d("TwilioInfo", "MemberMethods.getChannel => onSuccess")
                result.success(Mapper.channelToMap(channel))
            }

            override fun onError(errorInfo: ErrorInfo) {
                Log.d("TwilioInfo", "MemberMethods.getChannel => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getUserDescriptor(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        try {
            TwilioConversationsPlugin.chatClient?.getConversation(channelSid, object : CallbackListener<Conversation> {
                override fun onSuccess(channel: Conversation) {
                    Log.d("TwilioInfo", "MemberMethods.getUserDescriptor => onSuccess")
                    val member = channel.participantsList.find { it.identity == identity }
                    if (member != null) {
                        member.getAndSubscribeUser(object : CallbackListener<User> {
                            override fun onSuccess(userDescriptor: User) {
                                Log.d("TwilioInfo", "MemberMethods.getUserDescriptor (Member.getUserDescriptor) => onSuccess")
                                result.success(Mapper.userDescriptorToMap(userDescriptor))
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                Log.d("TwilioInfo", "ChannelsMethods.getUserDescriptor => getUserDescriptor onError: $errorInfo")
                                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                            }
                        })
                    } else {
                        return result.error("ERROR", "No member found on channel '$channelSid' with identity '$identity'", null)
                    }
                }

                override fun onError(errorInfo: ErrorInfo) {
                    Log.d("TwilioInfo", "ChannelsMethods.getUserDescriptor => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun getAndSubscribeUser(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val memberSid = call.argument<String>("memberSid")
                ?: return result.error("ERROR", "Missing 'memberSid'", null)

        try {
            TwilioConversationsPlugin.chatClient?.getConversation(channelSid, object : CallbackListener<Conversation> {
                override fun onSuccess(channel: Conversation) {
                    Log.d("TwilioInfo", "MemberMethods.getAndSubscribeUser => onSuccess")
                    val member = channel.participantsList.find { it.sid == memberSid }
                    if (member != null) {
                        member.getAndSubscribeUser(object : CallbackListener<User> {
                            override fun onSuccess(user: User) {
                                Log.d("TwilioInfo", "MemberMethods.getAndSubscribeUser (Member.getAndSubscribeUser) => onSuccess")
                                result.success(Mapper.userToMap(user))
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                Log.d("TwilioInfo", "ChannelsMethods.getAndSubscribeUser => getAndSubscribeUser onError: $errorInfo")
                                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                            }
                        })
                    } else {
                        return result.error("ERROR", "No member found on channel '$channelSid' with sid '$memberSid'", null)
                    }
                }

                override fun onError(errorInfo: ErrorInfo) {
                    Log.d("TwilioInfo", "ChannelsMethods.getAndSubscribeUser => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun setAttributes(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val memberSid = call.argument<String>("memberSid")
                ?: return result.error("ERROR", "Missing 'memberSid'", null)

        // Not erroring out because a nullable attributes is allowed to reset the Channel attributes.
        val attributes = call.argument<Map<String, Any>>("attributes")

        try {
            TwilioConversationsPlugin.chatClient?.getConversation(channelSid, object : CallbackListener<Conversation> {
                override fun onSuccess(channel: Conversation) {
                    Log.d("TwilioInfo", "MemberMethods.setAttributes => onSuccess")
                    val member = channel.participantsList.find { it.sid == memberSid }
                    if (member != null) {
                        member.setAttributes(Mapper.mapToAttributes(attributes) as Attributes, object : StatusListener {
                            override fun onSuccess() {
                                Log.d("TwilioInfo", "MemberMethods.setAttributes  (Channel.setAttributes) => onSuccess")
                                try {
                                    result.success(Mapper.attributesToMap(member.attributes))
                                } catch (err: JSONException) {
                                    return result.error("JSONException", err.message, null)
                                }
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                Log.d("TwilioInfo", "MemberMethods.setAttributes  (Channel.setAttributes) => onError: $errorInfo")
                                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                            }
                        })
                    } else {
                        return result.error("ERROR", "No member found on channel '$channelSid' with sid '$memberSid'", null)
                    }
                }

                override fun onError(errorInfo: ErrorInfo) {
                    Log.d("TwilioInfo", "MemberMethods.setAttributes => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }
}

File: flutter_twilio_conversations/ios/Classes/Methods/MemberMethods.swift
import Flutter
import Foundation
import TwilioConversationsClient

public class MemberMethods {
    public static func getUserDescriptor(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.conversation(
            withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHConversation?) in
                if result.isSuccessful, let channel = channel {
                    SwiftTwilioConversationsPlugin.debug("MemberMethods.getUserDescriptor => onSuccess")
                    if let member = channel.participant(withIdentity: identity) {
                        member.subscribedUser { (result: TCHResult, user: TCHUser?) in
                            if result.isSuccessful, let user = user {
                                flutterResult(Mapper.userDescriptorToDict(user))
                            } else {
                                flutterResult(FlutterError(
                                    code: "ERROR",
                                    message: "Error retrieving userDescriptor for member (identity: " +
                                    "'\(identity)') from channel (sid: \(channelSid)", details: nil))
                            }
                        }
                    } else {
                        flutterResult(FlutterError(code: "ERROR",
                                                   message: "Error retrieving member with identity '\(identity)'",
                            details: nil))
                    }
                } else {
                    SwiftTwilioConversationsPlugin.debug(
                        "MemberMethods.getUserDescriptor => onError: \(String(describing: result.error))")
                    flutterResult(FlutterError(
                        code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
                }
        })
    }

    public static func getAndSubscribeUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        guard let memberSid = arguments["memberSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'memberSid' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.conversation(
            withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHConversation?) in
                if result.isSuccessful, let channel = channel {
                    SwiftTwilioConversationsPlugin.debug("MemberMethods.getAndSubscribeUser => onSuccess")
                    if let member = channel.participant(withIdentity: memberSid) {
                        member.subscribedUser { (result: TCHResult, user: TCHUser?) in
                            if result.isSuccessful, let user = user {
                                flutterResult(Mapper.userToDict(user))
                            } else {
                                flutterResult(FlutterError(code: "ERROR",
                                    message: "Error subscribing to user with sid '\(memberSid)'",
                                    details: nil))
                            }
                        }
                    } else {
                        flutterResult(FlutterError(code: "ERROR",
                            message: "Error retrieving member with sid '\(memberSid)' " +
                            "from channel with sid \(channelSid)",
                            details: nil))
                    }
                } else {
                    if let error = result.error as NSError? {
                        SwiftTwilioConversationsPlugin.debug(
                            "MemberMethods.getAndSubscribeUser => onError: \(error)")
                        flutterResult(FlutterError(code: "ERROR",
                            message: "Error retrieving channel with sid or uniqueName '\(channelSid)'",
                            details: nil))
                    }
                }
        })
    }

    public static func setAttributes(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        guard let memberSid = arguments["memberSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'memberSid' parameter", details: nil))
        }

        guard let attributesDict = arguments["attributes"] as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'attributes' parameter", details: nil))
        }

        let attributes = TCHJsonAttributes.init(dictionary: attributesDict as [AnyHashable: Any])

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.conversation(
            withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHConversation?) in
                if result.isSuccessful, let channel = channel {
                    SwiftTwilioConversationsPlugin.debug("MemberMethods.setAttributes => onSuccess")
                    if let member = channel.participant(withIdentity: memberSid) {
                        member.setAttributes(attributes) { (result: TCHResult) in
                            if result.isSuccessful {
                                SwiftTwilioConversationsPlugin.debug(
                                    "MemberMethods.setAttributes (Member.setAttributes) => onSuccess")
                                flutterResult(Mapper.attributesToDict(attributes))
                            } else {
                                SwiftTwilioConversationsPlugin.debug(
                                    "MemberMethods.setAttributes (Member.setAttributes) => " +
                                    "onError: \(String(describing: result.error))")
                                flutterResult(FlutterError(code: "ERROR",
                                    message: "Error setting attributes for member with sid " +
                                    "'\(memberSid)'", details: nil))
                            }
                        }
                    }
                    flutterResult(Mapper.channelToDict(channel))
                } else {
                    if let error = result.error as NSError? {
                        SwiftTwilioConversationsPlugin.debug("MemberMethods.setAttributes => onError: \(error)")
                        flutterResult(FlutterError(code: "ERROR",
                            message: "Error retrieving channel with sid or uniqueName '\(channelSid)'", details: nil))
                    }
                }
        })
    }
}
